File: Reqnroll/Chapter16/16.2-ContextShortcutPattern/After/WIMP.Specs/Drivers/OrderingApiDriver.cs
using System.Net;

using WIMP.App.Models;
using WIMP.App.RestApi;
using WIMP.Specs.Support;

namespace WIMP.Specs.Drivers;

public class OrderingApiDriver(RestApiContext restApiContext)
{
    public TestAction<Order> PlaceOrder(PlaceOrderRequest placeOrderRequest) =>
        new LambdaAction<Order>("Place order", async () =>
            await restApiContext.ProcessRequest<Order>(
                "Place order", HttpMethod.Post, "/api/orders",
                placeOrderRequest, HttpStatusCode.Created));

    public TestAction<VoidReturn> ChangeDeliveryAddress(int orderNo, ChangeAddressRequest changeAddressRequest) =>
        new LambdaAction("Change delivery address", async () =>
            await restApiContext.ProcessRequest<VoidReturn>(
                "Change delivery address", HttpMethod.Put, $"/api/orders/{orderNo}/delivery-address",
                changeAddressRequest));
}

File: CucumberJVM/Chapter15/15.3-InfrastructureLayerPattern/After/src/test/java/com/wimp/app/specs/drivers/OrderingApiDriver.java
package com.wimp.app.specs.drivers;

import com.wimp.app.models.Order;
import com.wimp.app.restapi.PlaceOrderRequest;
import com.wimp.app.specs.support.TestActionFailedException;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.service.annotation.DeleteExchange;
import org.springframework.web.service.annotation.HttpExchange;
import org.springframework.web.service.annotation.PostExchange;

@Component
public class OrderingApiDriver {

    private final OrderingApiClient orderingApiClient;

    @HttpExchange("/api/orders")
    public interface OrderingApiClient {
        @PostExchange
        Order placeOrder(@RequestBody PlaceOrderRequest request);

        @DeleteExchange("/{orderNo}")
        void cancelOrder(@PathVariable int orderNo);
    }

    public OrderingApiDriver(OrderingApiClient orderingApiClient) {
        this.orderingApiClient = orderingApiClient;
    }

    public Order placeOrder(PlaceOrderRequest placeOrderRequest) {
        try {
            return orderingApiClient.placeOrder(placeOrderRequest);
        } catch (Exception ex) {
            throw new TestActionFailedException(ex.getMessage());
        }
    }

    public void cancelOrder(int orderNo) {
        try {
            orderingApiClient.cancelOrder(orderNo);
        } catch (Exception ex) {
            throw new TestActionFailedException(ex.getMessage());
        }
    }
}
